File: fortran/p_003.f90
! observasi ganteng:
! ide (1). All primes after 3 can be written in the form 6k ± 1
!           (the opposite doesn't necessarily hold, tho)
! ide (2). Other forms of numbers (6k, 6k + 2, 6k + 3, dan 6k + 4) are divisible by either 2 or 3
        !   Hence we can save up time and memory by immediately returning False when the input (n) is
        !   divisible by 2 or 3 instead of checking by dividing it one by one with numbers from 2 to n.
        !   We're left with numbers in the form of 6k ± 1. Mantap lur!
! ide (3). To check the primality of n, we only need to divide it with numbers up to sqrt(n)
! ide (4). Every number can be written in the form of (p_1)^(k_1) * (p_2)^(k_2) * ... * (p_n)^(k_n)
!           where p_i denotes the i-th prime factor of the number (sorted from smallest to largest)
!           and k_i is a natural number. (Well, pada dasarnya... faktorisasi prima)

! Program problem 3: menentukan faktor prima terbesar dari num = 600851475143
program problem3
    implicit none
    integer(16) :: num = 600851475143000000_16
    integer(8) :: thePrime = 1
    integer(8) :: maxPrime = 1
    integer(8) :: i

    ! Mencari faktor-faktor prima dari num
    do i = 5, int(num ** 0.5), 6
        if ((mod(num, i) .eq. 0) .or. mod(num,i+2) .eq. 0) then
            ! Periksa primalitas untuk bentuk 6k - 1
            if (isPrime(i)) then
                thePrime = i
            ! Periksa primalitas untuk bentuk 6k + 1
            else if (isPrime(i + 2)) then
                thePrime = i + 2
            end if

            ! terus membagi num dengan faktor prima yang ditemukan 
            ! hingga faktor tersebut tidak bisa membagi habis num lagi
            do while ((mod(num, thePrime) .eq. 0) .and. (num > 1))
                maxPrime = num
                num = num / thePrime
            end do
        end if
    end do

    ! OUTPUT, akhirnya!
    print *, maxPrime

    contains
    ! Fungsi isPrime(n) menguji primalitas n; mengembalikan .true. jika n prima, vice versa
    logical function isPrime(n)
        implicit none
        integer(8), intent(in) :: n
        integer :: i
        
        isPrime = .true.
        
        if (n <= 1) then ! 0, 1, dan bilangan negatif
            isPrime = .false.
        else if (n <= 3) then ! 2 dan 3
            isPrime = .true.
        else if ((mod(n, 2) .eq. 0) .or. (mod(n, 3) .eq. 0)) then ! kelipatan 2 dan 3
            isPrime = .false.
        else ! bentuk 6k + 1 atau 6k - 1
            do i = 5, int(n ** 0.5), 6
                if ((mod(n, i) .eq. 0) .or. (mod(n, i + 2) .eq. 0)) then
                    isPrime = .false.
                end if
            end do
        end if
    end function isPrime
end program problem3

File: python/p_003.py
"""
Largest Prime Factor
Problem 3

The prime factors of 13195 are 5, 7, 13, and 29. What is the largest
prime factor of the number 600851475143?

Observation:
Fact (1). All primes after 3 can be written in the form 6k ± 1
          (the opposite it doesn't necessarily hold, though)
Fact (2). Other forms of numbers (6k, 6k + 2, 6k + 3, dan 6k + 4) are divisible by either 2 or 3
          Hence we can save up time and memory by immediately returning False when the input (n) is
          divisible by 2 or 3 instead of checking by dividing it one by one with numbers from 2 to n.
          We're left with numbers in the form of 6k ± 1. 
Fact (3). To check the primality of n, we only need to divide it with numbers up to sqrt(n)
Fact (4). Every number can be written in the form of (p_1)^(k_1) * (p_2)^(k_2) * ... * (p_n)^(k_n)
          where p_i denotes the i-th prime factor of the number (sorted from smallest to largest)
          and k_i is a natural number.
Procedure:
Let num = 600851475143.
Based on Fact (4), we can do an iteration from i = 2 to i = sqrt(num). In each iteration, we check whether i is a prime. 
If yes, then divide num by i until it can't divide num evenly anymore (the number then becomes (p_2)^(k_2) * ... * (p_n)^(k_n)).
If not, immediately proceed to the next iteration.
Why do we only need to iterate up to sqrt(num)?
    - If num is a prime number then
"""

# With that in mind, we can define a prime checker function as follows:
def isPrime(n):
    if n <= 1:                      # 0, 1, and negative numbers aren't prime numbers
        return False
    elif n <= 3:                    # 2 and 3 are prime numbers
        return True
    elif n % 2 == 0 or n % 3 == 0:  # Other multiples of 2 or 3 aren't prime numbers
        return False
    for i in range(5, int(n ** 0.5) + 1, 6):  # Using Fact (1) and (3)
        if n % i == 0 or n % (i + 2) == 0:
            return False
    return True
num = 600851475143
maks = 2
for i in range(5, int(num ** 0.5 + 1), 6):
    if isPrime(i):
        while num % i == 0 and num > 1:
            maks = num
            num //= i
    if isPrime(i + 2):
        while num % (i + 2) == 0 and num > 1:
            maks = num
            num //= (i+2)
print(maks)